### ISSUE TITLE
refactor: POTENTIAL_ASSIGNMENTS don't agree with master/legacy

### ISSUE THREAD

**Issue by sbailey** [desihub/fiberassign#170]
@forero discovered this over in #165, but let's spin it into a new ticket:

The list of POTENTIAL_ASSIGNMENTS in the tsk_refactor branch is very different from master, e.g. from a run by @forero:
![51924407-fa324b00-23ba-11e9-8534-04f1e7ee1963](https://user-images.githubusercontent.com/218471/51934334-31ebc380-23b8-11e9-9ec1-1ec07e612624.png)

And from `/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor/` run by Stephen:
![51925297-bccebd00-23bc-11e9-8b21-7d91bfbe344a](https://user-images.githubusercontent.com/218471/51934338-33b58700-23b8-11e9-827b-92ac644f6140.png)

The sample with the big offset is from the bright tiles, but we don't yet know why (perhaps some target class isn't propagated into POTENTIAL_ASSIGNMENTS for master).  And even for dark and gray there should be better agreement between master and tsk_refactor since it is just a geometry calculation.

Perhaps the two codes have different ways of calculating whether the obsconditions allow that target on that tile, and whether or not targets should be propagated if they aren't allowed on the observing conditions for that tile?  e.g. if tsk_refactor is propagating all targets, and master is propagating only targets that are allowed to be observed under bright conditions, that would explain the big offset.  I don't have a good explanation for how the old code could find more potentials for dark and gray though.

**Comment by tskisner** [comment]
I don't have an obvious answer for this.  In the new code, the targets available for a given fiber should contain the sky, standards, and science targets *compatible with the tile/fiber obsconds*.  I can investigate more after #165.

**Comment by sbailey** [comment]
Update: There are two effects that I have found so far:

* refactor includes non-POS device locations, thus increasing the number of objects listed
* legacy includes targets that don't match the OBSCONDITIONS of the tile, thus increasing the number of objects listed

I'm still digging for how closely the two agree after correcting for these two effects.

**Comment by sbailey** [comment]
Two more sources of discrepancy:
* refactor includes sky targets while legacy does not
* refactor includes standards that are not also MWS targets, while legacy does not

After accounting for all those effects, the list of unique TARGETIDs matches between master and tsk_refactor for every tile, however, master still has more claimed potential assignment combinations than tsk_refactor.  Chasing that next...

**Comment by sbailey** [comment]
The last remaining discrepancy was that master includes duplicate entries.  Summarizing the differences between tsk_refactor and legacy (current master):
* refactor includes non-POS device locations, thus increasing the number of objects listed
* legacy includes targets that don't match the OBSCONDITIONS of the tile, thus increasing the number of objects listed
* legacy includes only science targets, while refactor includes science, std, and sky targets
* legacy has duplicate entries

I suggest that we update tsk_refactor to only include entries for POS and ETC but not the non-positioner locations (FIF, NON, ...)  Otherwise I think tsk_refactor is doing what we want and it probably isn't worth it to update legacy to match.

For the record, code to trim the two POTENTIAL_ASSIGNMENT lists to match:
```python
import numpy as np
import fitsio
import desimodel.io
from desitarget.targetmask import desi_mask

tileid = 26819
basedir = '/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor'

p1 = fitsio.read(basedir+'/master/tile-{:05d}.fits'.format(tileid), 'POTENTIAL_ASSIGNMENTS')
p2 = fitsio.read(basedir+'/output/tile-{:05d}.fits'.format(tileid), 'POTENTIAL_ASSIGNMENTS')

tiles = fitsio.read(basedir + '/tiles-lite.fits')
tileobscond = tiles['OBSCONDITIONS'][tiles['TILEID']==tileid][0]

deviceloc = desimodel.io.load_deviceloc()
posloc = deviceloc['LOCATION'][deviceloc['DEVICE_TYPE'] == 'POS']

mtl = fitsio.read(basedir+'/mtl.fits', 'MTL')
skies = fitsio.read(basedir+'/skies-dr7.1-0.27.0-lite.fits', 'SKIES')

#- Trim p1 to only targets matching the obsconditions for this tile
ii = np.in1d(mtl['TARGETID'], p1['TARGETID'])
jj = (mtl['OBSCONDITIONS'][ii] & tileobscond) != 0
kk = np.in1d(p1['TARGETID'], mtl['TARGETID'][ii][jj])
p1 = p1[kk]

#- remove p1 PRIORITY=0 targets
jj = (mtl['PRIORITY'][ii] > 0)
kk = np.in1d(p1['TARGETID'], mtl['TARGETID'][ii][jj])
p1 = p1[kk]

#- remove duplicate p1 entries
p1 = np.unique(p1)
    
#- Trim p2 to only POS devices
keep = np.in1d(p2['LOCATION'], posloc)
p2 = p2[keep]

#- remove p2 sky targets
keep = ~np.in1d(p2['TARGETID'], skies['TARGETID'])
p2 = p2[keep]

#- Remove p2 STD-only targets
pureSTD = (mtl['DESI_TARGET'] & desi_mask.mask('STD_FAINT|STD_BRIGHT|STD_WD')) == mtl['DESI_TARGET'] 
keep = ~np.in1d(p2['TARGETID'], mtl['TARGETID'][pureSTD])
p2 = p2[keep]

assert np.all(np.in1d(p1, p2))
assert np.all(np.in1d(p2, p1))
```

**Comment by tskisner** [comment]
I'll update the device types to include only POS and ETC and close this when it is pushed

**Comment by tskisner** [comment]
Should be fixed in 658c6d3bc9f44f83d1b686c901339030218db3d5
